src/scale_impls/decode.rs

```diff
@@ -23,424 +23,473 @@ pub use scale_decode::visitor::DecodeError;
-		self,
-		value: i64,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		self.visit_i128(value as i128, type_id)
-	}
-	fn visit_i128(
-		self,
-		value: i128,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		Ok(Value::i128(value as i128).map_context(|_| type_id.into()))
-	}
-	fn visit_i256(
-		self,
-		value: &[u8; 32],
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		Ok(Value { value: ValueDef::Primitive(Primitive::U256(*value)), context: type_id.into() })
-	}
-	fn visit_sequence(
-		self,
-		value: &mut scale_decode::visitor::Sequence,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		to_unnamed_composite!(value, type_id)
-	}
-	fn visit_tuple(
-		self,
-		value: &mut scale_decode::visitor::Tuple,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		to_unnamed_composite!(value, type_id)
-	}
-	fn visit_array(
-		self,
-		value: &mut scale_decode::visitor::Array,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		to_unnamed_composite!(value, type_id)
-	}
-	fn visit_bitsequence(
-		self,
-		value: &mut scale_decode::visitor::BitSequence,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		let bits: Result<_, _> = value.decode()?.collect();
-		Ok(Value { value: ValueDef::BitSequence(bits?), context: type_id.into() })
-	}
-	fn visit_str(
-		self,
-		value: scale_decode::visitor::Str,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		Ok(Value::string(value.as_str()?).map_context(|_| type_id.into()))
-	}
-	fn visit_variant(
-		self,
-		value: &mut scale_decode::visitor::Variant,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		let values = visit_composite(value.fields())?;
-		Ok(Value {
-			value: ValueDef::Variant(Variant { name: value.name().to_owned(), values }),
-			context: type_id.into(),
-		})
-	}
-	fn visit_composite(
-		self,
-		value: &mut scale_decode::visitor::Composite,
-		type_id: scale_decode::visitor::TypeId,
-	) -> Result<Self::Value, Self::Error> {
-		Ok(Value { value: ValueDef::Composite(visit_composite(value)?), context: type_id.into() })
-	}
+// We can't implement this on `Value<TypeId>` because we have no TypeId to assign to the value.
+impl scale_decode::DecodeAsFields for Composite<TypeId> {
+    fn decode_as_fields(
+        input: &mut &[u8],
+        fields: &[scale_decode::PortableField],
+        types: &PortableRegistry,
+    ) -> Result<Self, scale_decode::Error> {
+        // Build a Composite type to pass to a one-off visitor:
+        let path = Default::default();
+        let mut composite =
+            scale_decode::visitor::types::Composite::new(input, &path, fields, types);
+        // Decode into a Composite value from this:
+        let val = visit_composite(&mut composite)?;
```

Comment:
nit: Wouldn't we want to keep consistency with the `scale_decode` pattern where we advance the remaining data regardless of the result here? Meaning do we want to update the input cursor only if`visit_comoposite` is successful, or all the time (as scale_decode)? 🤔

Reply:
scale_decode has two impls of `decode_as_fields`, one for tuples and one in the proc macro. The tuple one doesn't advance at all ever (I'm going to fix that and add a tuple to the test there!), and the proc macro looks like it does the same as this; advances if everything is ok but not if error here.

I think it should def advance when things are OK. I'm not as bothered on error (because we control the visitor impls that are used anyway) but let's be defensive and fix this here and in scale- decode (I'll do that now!)